dataquality/utils/seq2seq.py

```diff
@@ -4,8 +4,152 @@
+        - The sample is represented as a List of Lists per token.
+        - For each token, we store a fixed length (k) list of tuples for each of
+        the top-k predicted tokens - (token_string, logprob).
+
+    Parameters:
+    -----------
+    sample_logprobs: np.ndarray of shape - [seq_len, Vocab size]
+    top_indices: np.ndarray of shape - [seq_len, k]
+    tokenizer: PreTrainedTokenizerFast
+
+    Return:
+    -------
+    top_logprobs: List[List[Tuple[str, float]]
+        len(top_logprobs) == sample_logprobs.shape[0] == num_tokens
+        len(top_logprobs[i]) == TOP_K
+    """
+    # Extract top_k logprob indices - shape = [seq_len, k]
+    sample_top_logprobs = np.take_along_axis(sample_logprobs, top_indices, axis=-1)
+
+    # Generate top_k (string, logprob) token by token
+    top_logprobs: List[List[Tuple[str, float]]] = []
+    for token_top_ids, token_top_logprobs in zip(top_indices, sample_top_logprobs):
+        # List of Tuple[str, int] --> (token string, logprob)
+        token_top_logprobs_mapping = []
+        # Loop over the top_k predictions for the given token position
+        for pred_token_id, logprob in zip(token_top_ids, token_top_logprobs):
+            str_token = tokenizer.decode(pred_token_id)
+            token_top_logprobs_mapping.append((str_token, logprob))
+
+        top_logprobs.append(token_top_logprobs_mapping)
+
+    return top_logprobs
+
+
+def process_sample_logprobs(
+    sample_logprobs: np.ndarray,
+    sample_labels: np.ndarray,
+    sample_top_indices: np.ndarray,
+    tokenizer: PreTrainedTokenizerFast,
+) -> Tuple[np.ndarray, List[List[Tuple[str, float]]]]:
+    """Extract label_logprobs and top_logprobs
+
+    Whether the labels are GT target labels or generated labels, the
+    process is identical. Extract the per token probability assigned to the
+    token label and the top-k logprobs.
+
+    Preconditions:
+        - We assume that all inputs have been stripped of any padding tokens!
+
+    Parameters:
+    -----------
+    sample_logprobs: np.ndarray of shape - [seq_len, vocab_size]
+        Per-token logprobs for a single sample
+    sample_labels: np.ndarray of shape - [seq_len]
+        Per-token lables for the sample. As a pre-condition
+        we assume that this is a 1D tensor with length seq_len.
+        This is important for extracting logprobs
+    sample_top_indices: np.ndarray of shape - [seq_len, TOP_K]
+        Top_K logprob indices for each token. Note that these
+        are not in order.
+    tokenizer: PreTrainedTokenizerFast
+        Tokenizer used by the model
+
+    Returns:
+    --------
+    label_logprobs: np.ndarray of shape [seq_len]
+        Label logprobs for each token in the sample
+    top_logprobs: List[Dict[str, float]]
+        List of top-k (str) predictions + corresponding logprobs
+    """
+    # Ensure final shape - [len(labels), 1]
+    if sample_labels.ndim != 1:
+        raise GalileoException("Expects sample_labels to be a 1D array")
```

Comment:
```suggestion
    if sample_labels.ndim != 1:
        raise GalileoException(f"Invalid shape {sample_lables.shape}, process_sample_logprobs expects sample_labels to be a 1D array")
```

you'll have to format ^